src/inmanta/parser/plyInmantaParser.py

```diff
@@ -731,40 +733,52 @@ def p_constant_rstring(p: YaccProduction) -> None:
 
 def p_constant_mls(p: YaccProduction) -> None:
     "constant : mls"
-    p[0] = get_string_ast_node(p[1], p[1].location)
+    p[0] = get_string_ast_node(p[1], True)
     attach_from_string(p)
 
 
 format_regex = r"""({{\s*([\.A-Za-z0-9_-]+)\s*}})"""
 format_regex_compiled = re.compile(format_regex, re.MULTILINE | re.DOTALL)
 
 
-def get_string_ast_node(string: LocatableString, location: Location) -> Union[Literal, StringFormat]:
-    match_obj = format_regex_compiled.findall(str(string))
-    if len(match_obj) == 0:
+def get_string_ast_node(string: LocatableString, mls: boolean) -> Union[Literal, StringFormat]:
+    matches = [[str(string)[m.start() : m.end()], m.start(), m.end()] for m in format_regex_compiled.finditer(str(string))]
+    if len(matches) == 0:
         return Literal(str(string))
-    return create_string_format(string, match_obj, location)
 
-
-def create_string_format(format_string: LocatableString, variables: List[List[str]], location: Location) -> StringFormat:
+    offset = 3 if mls else 1  # len(""")  or len(') or len(")
+    init: List[str]
+    match: List[str]
+    locatable_matches: List(LocatableString) = []
+    for obj in matches:
+        init, match, _ = (str(string)[start:end].splitlines() for start, end in pairwise((0, obj[1], obj[2], len(obj))))
+        start_line: int = string.lnr + len(init) - 1
+        end_line: int = start_line + len(match) - 1
+        len_last_init_line: int = len(init[-1]) if init else 0
+        start_char: int = len(init[-1]) + offset + string.start if len(init) == 1 else len_last_init_line + 1
+        end_char: int = len(match[-1]) + start_char - 1 if len(match) == 1 else len(match[-1])
+        range: Range = Range(string.location.file, start_line, start_char, end_line, end_char)
+        locatable_string = LocatableString("".join(match), range, string.lexpos, string.namespace)
```

Comment:
`"".join(match)` is a simplification of the actual format string. At first I thought it wouldn't matter at all but when looking into `StringFormat.execute` it seems that it expects the variables as passed in its constructor to appear exactly like that in the string. Meaning that if any newlines occur between the `{{` and `}}`, this won't work. `"\n".join(match)` would not be sufficient either because of potentially different line endings (`\r\n` for example).

So, `create_string_format` requires the exact string (`obj[0]` I think). There are a couple of different approaches of course. My suggestion would be to combine this with my comment regarding using the `Match` object (can't link there for some reason). You could then easily have this whole loop construct a `LocatableString` just for the variable part, and then pass  a tuple of that and the exact substring (`match.group(0)`) to `create_string_format`.

Reply:
So by exact string, It would be the string before it gets "splitlined"? can't I just use that as name for the `LocatableString`? and use that in `create_string_format`.
like this:
```
 for obj in matches:
        init_string, match_string = (str(string_ast)[start:end] for start, end in pairwise((0, obj.start(), obj.end())))
        init: List[str] = (init_string + "\n").splitlines()     # lines before the match
        match: List[str] = match_string.splitlines()       # lines matching the string format
        ....
        locatable_string = LocatableString(match_string, range, string_ast.lexpos, string_ast.namespace)
        locatable_matches.append(locatable_string)
 return create_string_format(string_ast, locatable_matches)
        
```

Reply:
That is one of the options. I think the main advantage of the approach I suggested would be that we immediately have the `Range` for the actual attribute reference instead of the `Range` including `{{ }}` and any whitespace in between. I think that could lead to a cleaner code overall because it would increase accuracy while keeping the code very simple, only at the cost of extending the method to accept a tuple. But if you disagree I'm not against the approach you suggest. Do keep in mind thought that it will require some more advanced logic in `create_string_format` if we want to determine the correct `Range` for the `Reference` and `AttributeReference` values.